Kernel/Devices/KCOVDevice.cpp

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2021, Patrick Meyer <git@the-space.agency>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Assertions.h>
+#include <AK/Format.h>
+#include <AK/NonnullOwnPtr.h>
+#include <Kernel/Devices/KCOVDevice.h>
+#include <Kernel/Devices/KCOVInstance.h>
+#include <Kernel/FileSystem/FileDescription.h>
+#include <Kernel/Sections.h>
+#include <Kernel/VM/AnonymousVMObject.h>
+#include <LibC/sys/ioctl_numbers.h>
+
+#include <Kernel/Panic.h>
+
+namespace Kernel {
+
+HashMap<ProcessID, KCOVInstance*>* KCOVDevice::proc_instance;
+HashMap<ThreadID, KCOVInstance*>* KCOVDevice::thread_instance;
+
+UNMAP_AFTER_INIT NonnullRefPtr<KCOVDevice> KCOVDevice::must_create()
+{
+    return adopt_ref_if_nonnull(new KCOVDevice).release_nonnull();
+}
+
+UNMAP_AFTER_INIT KCOVDevice::KCOVDevice()
+    : BlockDevice(30, 0)
+{
+    proc_instance = new HashMap<ProcessID, KCOVInstance*>();
+    thread_instance = new HashMap<ThreadID, KCOVInstance*>();
+    dbgln("KCOVDevice created");
+}
+
+void KCOVDevice::free_thread()
+{
+    auto thread = Thread::current();
+    auto tid = (*thread).tid();
+
+    auto maybe_kcov_instance = thread_instance->get(tid);
+    if (!maybe_kcov_instance.has_value())
+        return;
+
+    auto kcov_instance = maybe_kcov_instance.value();
+    VERIFY(kcov_instance->state == KCOVInstance::TRACING);
+    kcov_instance->state = KCOVInstance::OPENED;
+    thread_instance->remove(tid);
+}
+
+void KCOVDevice::free_process()
+{
+    auto process = Process::current();
+    auto pid = (*process).pid();
+
+    auto maybe_kcov_instance = proc_instance->get(pid);
+    if (!maybe_kcov_instance.has_value())
+        return;
+
+    auto kcov_instance = maybe_kcov_instance.value();
+    VERIFY(kcov_instance->state == KCOVInstance::OPENED);
+    kcov_instance->state = KCOVInstance::UNUSED;
+    proc_instance->remove(pid);
+}
+
+KResultOr<NonnullRefPtr<FileDescription>> KCOVDevice::open(int options)
+{
+    auto process = Process::current();
+    if (!process) {
+        PANIC("Couldn't get caller process while initializing KCOV instance :^(");
+    }
```

Comment:
```suggestion
    if (!process)
        PANIC("Couldn't get caller process while initializing KCOV instance :^(");
```

You can omit curly braces for single line blocks.

Reply:
ack